### DISCUSSION TITLE
TM1PY and IBM Cloud Rich Tier

### DISCUSSION THREAD

**Discussion by harveyca307** [cubewise-code/tm1py#754] [General]
Working on a client deliverable and running into an odd issue.  I am creating a programmatic interface between PA and an outside project management tool using RESTful APIs.  I've written and tested the Python and compiled to an EXE using PYInstaller, as I know of no other way to run Python on the cloud server.  

The programs runs exactly like it is supposed to from Command Line or Powershell; however, when I issue an ExecuteCommand from TI it just hangs.  I've waited an appropriate amount of time for the TI to finish before cancelling the thread.  Anyone else seeing an issue like this?  

My code uses TM1Py to write values to the PA server using cube.cells.write_values, and uses the external API to push PA projects to them...

**Comment by MariusWirtz** [comment]
Does it hang if you call the `ExecuteCommand` TI function with the second argument as 0?

**Reply by harveyca307** [reply]
I haven't tried with the second as argument as 0.  Let me try

**Reply by harveyca307** [reply]
It did not hang; however it also didn't do anything.

**Reply by harveyca307** [reply]
Interestingly enough as I was looking at PythonAnywhere, I started to think about how it would get the password for the non-interactive user.  This lead me to think that maybe the PA service user wouldn't have access to the Windows Keyring (where I was going to store the password).  Turns out that was correct.  I hardcoded the non-interactive user password in the compiled code and it worked.  Now I need to come up with a better way of storing/using that password...

As always, you're awesome Marius!

**Reply by MariusWirtz** [reply]
You can run a separate python script that uses keyring to store a password for the PA service user ;)

That script you run once initially and then the other scripts can retrieve the password safely with keyring.

**Comment by MariusWirtz** [comment]
On a side note, there are alternative ways to deploy TM1py scripts for TM1 in the IBM cloud.
For instance, you could run your script at [pythonanywhere](https://www.pythonanywhere.com/).

Disclaimer: I haven't used **pythonanywhere** personally for production deployments.
